### PR TITLE
Add nlp blogpost

### LINKED ISSUE #15: Blogpost about NLP Results

### PR DESCRIPTION
**Pull request by ehoelzl**
Closes #15

### CODE REVIEW

File: _posts/2020-10-02-nlp-translation.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+excerpt_separator: <!--more-->
+---
+
+Natural Language Processing (NLP) has gained a lot of interest in the Machine Learning community. 
+This increasing attention towards the subject may come from the fascination of teaching machines to understand and assimilate human language, 
+and use them as tools to complement and facilitate our everyday lives. 
+
+Machine translation is one branch of NLP, and consists of having automated model capable of translating text from one language to another in a few seconds. 
+
+In this blog post, we analyze the performance increase brought by distribution for two different Machine Translation models: an LSTM variant (GNMT) and an attention
```
Comment (martinjaggi) [inline _posts/2020-10-02-nlp-translation.md]: distributed training

File: _posts/2020-10-02-nlp-translation.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+title: NLP Translation tasks, results discussion
+author: e_hoelzl
+published: true
+tags: [performance, results]
+excerpt_separator: <!--more-->
+---
+
+Natural Language Processing (NLP) has gained a lot of interest in the Machine Learning community. 
+This increasing attention towards the subject may come from the fascination of teaching machines to understand and assimilate human language, 
+and use them as tools to complement and facilitate our everyday lives.
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md]: I like the second part of this paragraph, but not the first line. "has gained a lot of interest" sounds to me like NLP is a recent thing. I personally see it as a very established area in ML. Maybe you could start with something like "The popularity and relevance of NLP may come from ..." or so. Or just "The fascinating field of NLP enables us to ..." (and then join it with the next paragraph)

File: _posts/2020-10-02-nlp-translation.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+published: true
+tags: [performance, results]
+excerpt_separator: <!--more-->
+---
+
+Natural Language Processing (NLP) has gained a lot of interest in the Machine Learning community. 
+This increasing attention towards the subject may come from the fascination of teaching machines to understand and assimilate human language, 
+and use them as tools to complement and facilitate our everyday lives. 
+
+Machine translation is one branch of NLP, and consists of having automated model capable of translating text from one language to another in a few seconds.
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md]: 'a few seconds' sounds slow to me. Maybe you can say something more vague that means 'fast'

File: _posts/2020-10-02-nlp-translation.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+excerpt_separator: <!--more-->
+---
+
+Natural Language Processing (NLP) has gained a lot of interest in the Machine Learning community. 
+This increasing attention towards the subject may come from the fascination of teaching machines to understand and assimilate human language, 
+and use them as tools to complement and facilitate our everyday lives. 
+
+Machine translation is one branch of NLP, and consists of having automated model capable of translating text from one language to another in a few seconds. 
+
+In this blog post, we analyze the performance increase brought by distributed training for two different Machine Translation models: an LSTM variant (GNMT) and an attention
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md]: suggestion: "In this post, we analyze how distributed learning improves the training time in two machine translation models:

File: _posts/2020-10-02-nlp-translation.md [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+Machine translation is one branch of NLP, and consists of having automated model capable of translating text from one language to another in a few seconds. 
+
+In this blog post, we analyze the performance increase brought by distributed training for two different Machine Translation models: an LSTM variant (GNMT) and an attention 
+based model (Transformer).
+
+<!--more-->
+
+Those models present two main limitations that makes training very time consuming:
+ - They need millions of data points to reach acceptable performance.
+ - Models are quite large, and computations take significantly more time.
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md]: This sounds imprecise to me. "quite large compared to what?", "more time than what?"

File: _posts/2020-10-02-nlp-translation.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+Each of those problems can be solved using distribution:
+ - Distribute the data on multiple machines (data-parallel)
+ - Distributed computations on multiple cores (compute-parallel)
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md]: maybe specify 'computations for one datapoint', otherwise splitting the data automatically also splits the total computation

File: _posts/2020-10-02-nlp-translation.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+ 
+However the second approach requires the models to be parallelizable. We have only used data-parallel distribution for those tasks, but hope to combine it
+with parallel computation in the future.
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md]: I'm not so fond of the terminology 'parallel computation' here. Data-parallel training also involves parallel computation. 

Maybe the story could go like 
"Based on these limitations, we can divide processing of datapoints over multiple workers, or even subdivide the computations required to process a single datapoint. In our experiments, we focus on dividing the data (data-parallel). We plan to extend these results to model-parallel training in the future.

File: _posts/2020-10-02-nlp-translation.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+The attention module is based on the one described in {% cite bahdanau2014neural %}
+
+In our implementation, the encoder and decoder have each 4 stacked LSTM layers with residual connections, and hidden sizes of 1024. 
+This gives a model with a total of 160,671,297 trainable parameters. 
+
+### Transformer
+
+This model was first published in {% cite attention %}, and aims at completely disregarding recurrence and relying entirely on self-attention 
+mechanisms to perform sequence modelling and translation problems. 
+Such a structure allows for better parallelization of training on multiple GPUs, and can reach significantly better performance than comparable models.
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md]: If this is about model-parallel training, maybe you should leave it out to avoid confusion.

File: _posts/2020-10-02-nlp-translation.md [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+From this, we can compute $$NLLLoss(\mathbf{\hat{y}}, \mathbf{y})$$ where $$\mathbf{\hat{y}}$$ is the model output and $\mathbf{y}$ is the target. 
+
+
+$$ Smooth Loss = -mean (log softmax(\mathbf{\hat{y}})) $$
+$$ Loss = confidence * NLLLoss + smoothing * SmoothLoss $$
+
+Where $$confidence = 1 - smoothing$$. The smoothing is set to a value of 0.1 for both tasks.
+
+### Optimization
+As we have seen above, both models have a very high number of parameters to train. This can be an issue when using GPUs, as the model needs to fit in memory.
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md]: Maybe specify how big the models are, and how much GPU memory you need exactly to train this model. Does it not fit in 32-bit floats?

File: _posts/2020-10-02-nlp-translation.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+Also, back-propagation requires the memory to be at least twice the size of the model for it to work in memory. For that, instead of using regular precision, we used
+mixed-precision training, where most computations are done in `Float16`. We use a centralized version of `Adam`, where gradients are aggregated amongst all workers before
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md]: I was confused for a moment about the word 'centralized'. It made me think 'single worker'. Suggestion: "We use synchronous data-parallel training with Adam on gradients that are averaged between all workers."

File: _posts/2020-10-02-nlp-translation.md [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+mixed-precision training, where most computations are done in `Float16`. We use a centralized version of `Adam`, where gradients are aggregated amongst all workers before 
+updating weights.
+
+### Datasets
+Both tasks use the same test set, but are trained on slightly different data sets:
+- The LSTM is trained on the English to German World Machine Translation 16 (WMT16) dataset, comprising of 3,975,116 translated sentences
+- The Transformer is trained on the English to German World Machine Translation 17 (WMT17) dataset, comprising of 4,590,101.
+ 
+
+More details on both tasks can be found on our [documentation](https://mlbench.readthedocs.io/en/latest/benchmark-tasks.html#task-4-machine-translation).
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md]: *in* our documentation

File: _posts/2020-10-02-nlp-translation.md [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+
+The goal for both models is determined by the Bilingual Evaluation Understudy Score (BLEU):
+- The LSTM task stops when reaching a BLEU Score of 24.0
+- The Transformer task stops when reaching a BLEU score of 25.0
+
+The models are trained on 1,2,4,8 and 16 workers, and all step times are precisely measured to obtain an accurate speed up quantification.
+Speedups are computed with respect to the 1 worker case, and are intended to illustrate the distributive capabilities of the task.
+
+### Overall Speedups
+
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md:142]: Are you splitting a fixed batch size over more and more workers, or increasing the batch size?

File: _posts/2020-10-02-nlp-translation.md [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+$$ Smooth Loss = -mean (log softmax(\mathbf{\hat{y}})) $$
+$$ Loss = confidence * NLLLoss + smoothing * SmoothLoss $$
+
+Where $$confidence = 1 - smoothing$$. The smoothing is set to a value of 0.1 for both tasks.
+
+### Optimization
+As we have seen above, both models have a very high number of parameters to train. This can be an issue when using GPUs, as the model needs to fit in memory.
+Also, back-propagation requires the memory to be at least twice the size of the model for it to work in memory. For that, instead of using regular precision, we used
+mixed-precision training, where most computations are done in `Float16`. We use a centralized version of `Adam`, where gradients are aggregated amongst all workers before 
+updating weights.
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md]: Are you using something like 'DistributedDataParallel' that overlaps communication with computation?

File: _posts/2020-10-02-nlp-translation.md [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+  ![test]({{ site.baseurl }}public/images/blog/2020-10-02-nlp-translation/task4b_speedup.png)
+  *Transformer Speedups*
+</a>
+
+The left graph shows the absolute speed ups with respect to one worker, and the right one omits
+communication times from the speed up. This allows us to better see the effect of communication.
+
+
+A few interesting points:
+- Overall speedups follows a $$ log_{2}(n) $$, with `n = num_workers`, while compute are roughly linear.
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md]: I don't feel so confident about the log-2 relation. I don't think that would extrapolate. Maybe you can replace this with a less mathematical statement.

File: _posts/2020-10-02-nlp-translation.md [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+- Scaling the number of compute nodes gives nearly perfect scaling for both tasks (right plot)
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md:164]: This is not so surprising to me. I would focus on the other results.

File: _posts/2020-10-02-nlp-translation.md [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+- Using more powerful communication hardware (e.g. Tesla V100) will positively affect speedups.
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md]: How did you conclude this from the graphs above?

File: _posts/2020-10-02-nlp-translation.md [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+The models are trained on 1,2,4,8 and 16 workers, and all step times are precisely measured to obtain an accurate speed up quantification.
+Speedups are computed with respect to the 1 worker case, and are intended to illustrate the distributive capabilities of the task.
+
+### Overall Speedups
+
+The graphs below show the time speedups for the LSTM model and Transformer model (respectively). 
+
+<a href="{{ site.baseurl }}public/images/blog/2020-10-02-nlp-translation/task4a_speedup.png" data-lightbox="task4a_speedups" data-title="Speedups for GNMT">
+  ![test]({{ site.baseurl }}public/images/blog/2020-10-02-nlp-translation/task4a_speedup.png)
+  *GNMT Speedups*
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md:146]: With the spacing on the blog post, the label "GNMT Speedups" is visually closer to the graphs below than above. Maybe you can add some spacing below, or move the labels to above the plots.

File: _posts/2020-10-02-nlp-translation.md [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+- Overall speedups follows a $$ log_{2}(n) $$, with `n = num_workers`, while compute are roughly linear.
+- Scaling the number of compute nodes gives nearly perfect scaling for both tasks (right plot)
+- Using more powerful communication hardware (e.g. Tesla V100) will positively affect speedups.
+
+As the distribution level increases, we can see that communication becomes more and more heavy, and attenuates speedups quite significantly.
+
+### Step times
+
+The next figures show the total time spent in each step of training. 
+
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md:173]: I think it would be good to have a shared y axis here, so we can compare the times of various operations.

Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md:173]: It's not completely clear to me which graphs 'include' which others. Maybe you could add some kind of diagram that shows something like
```
computation = forward + backward
total = computation + communication
```
for each of the labels used.

File: _posts/2020-10-02-nlp-translation.md [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+  *Step times for GNMT*
+</a>
+
+<a href="{{ site.baseurl }}public/images/blog/2020-10-02-nlp-translation/task4b_times.png" data-lightbox="task4b_times" data-title="Step times for Transformer">
+  ![test]({{ site.baseurl }}public/images/blog/2020-10-02-nlp-translation/task4b_times.png)
+  *Step times for Transformer*
+</a>
+
+As expected, we can see that compute steps take less time as we increase the number of nodes,
+while communication increasingly takes more and more time, following a logarithmic path. Interestingly, the Transformer model's communication times quickly reach a plateau
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md]: I would skip the 'logarithmic' here. Not sure if it would extrapolate.

File: _posts/2020-10-02-nlp-translation.md [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+Compared to the LSTM model, the communication time ratio is slighly lower, but follows a similar path. For 16 workers, it reaches above 60%. However, the price index has a different shape:
+the observed valley is missing, and the indices are decreasing as we add workers. This suggests a very good performance increase, with a lower price increase. The best configuration 
+according to this index is with 8 workers, but the 16 worker case still has very impressive advantages.
+
+-----
+
+## References
+
+
+{% bibliography --cited %}
```
Comment (tvogels) [inline _posts/2020-10-02-nlp-translation.md:234]: Additionally to the results that are already here, I would find it interesting to compare the computation time to communication time ratio at a fixed 'distribution level' between transformers and LSTMs. I guess that quantity determines how well it scales.

**Review by tvogels — COMMENTED** [review]
Hi @ehoelzl, @martinjaggi asked me to go over the blog post and leave a review.
Interesting results. I also like the overall tone of the blog post.
My comments are a mix of reformulation suggestions to make things more concise, terminology confusion, and a thing or two about the presentation of the results. Hope this is helpful.

File: _posts/2020-10-02-nlp-translation.md (hunk 9 above)
Comment (ehoelzl) [reply _posts/2020-10-02-nlp-translation.md]: Not with the used batch size and GPUs (16GB memory)

File: _posts/2020-10-02-nlp-translation.md (hunk 10 above)
Comment (ehoelzl) [reply _posts/2020-10-02-nlp-translation.md]: They are not always averaged, sometimes the aggregation is different than just an average.

File: _posts/2020-10-02-nlp-translation.md (hunk 12 above)
Comment (ehoelzl) [reply _posts/2020-10-02-nlp-translation.md:142]: Splitting a fixed batch size over more workers. Global (sum of all workers) batch size is always the same, otherwise we would need to change learning rates and other parameters

File: _posts/2020-10-02-nlp-translation.md (hunk 13 above)
Comment (ehoelzl) [reply _posts/2020-10-02-nlp-translation.md]: Nope, tensors are first aggregated synchronously and then all computation is done (or vice-versa). We currently don't overlap communication with computation

File: _posts/2020-10-02-nlp-translation.md (hunk 18 above)
Comment (ehoelzl) [reply _posts/2020-10-02-nlp-translation.md:173]: The problem is that we wouldn't be able to see the improvements on some plots. (fwd pass, communication and optimizer)

File: _posts/2020-10-02-nlp-translation.md (hunk 16 above)
Comment (ehoelzl) [reply _posts/2020-10-02-nlp-translation.md]: We currently have around 10Gbps. V100 offer up to 300

**Comment by ehoelzl** [comment]
@martinjaggi @tvogels Could you review the updates please ? Thanks

File: _posts/2020-10-02-nlp-translation.md [hunk 21]
```diff
@@ -181,45 +182,49 @@ The next figures show the total time spent in each step of training.
...
+Here, *All Reduce* aggregation of gradients is performed before the optimization step, which yields a lot of exchanged messages. It would be interesting to see how the time spent communicating
+tensors can be reduced by using a more advanced aggregation technique (e.g. sharing with neighbors in a pre-defined topology)
 
 
 #### Transformer
-<a href="{{ site.baseurl }}public/images/blog/2020-10-02-nlp-translation/task4b_loss_ratio_prices.png" data-lightbox="task4b_loss_ratio_prices" data-title="Step times for Transformer">
-  ![test]({{ site.baseurl }}public/images/blog/2020-10-02-nlp-translation/task4b_loss_ratio_prices.png)
-  *Step times for Transformer*
+<a href="{{ site.baseurl }}public/images/blog/2020-10-02-nlp-translation/task4b_step_shares.png" data-lightbox="task4b_step_shares" data-title="Step shares for Transformer">
+  *Step shares for Transformer*
```
Comment (martinjaggi) [inline _posts/2020-10-02-nlp-translation.md:217]: is a 'step share' a communication step? maybe better to call it that in the text?

**Comment by martinjaggi** [comment]
very nice!
i left minor cosmetic comments
